staging/src/k8s.io/client-go/examples/controller-ref-manager/controller/ref_manager.go

```diff
@@ -0,0 +1,142 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	basecontroller "k8s.io/client-go/tools/controller"
+)
+
+// MyController is simple controller mock
+type MyController struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+}
+
+// DeploymentControllerRefManager manage deployments owner references
+type DeploymentControllerRefManager struct {
+	basecontroller.BaseControllerRefManager
+	controllerKind schema.GroupVersionKind
+	client         kubernetes.Interface
+}
+
+// NewDeploymentControllerRefManager returns a DeploymentControllerRefManager that
+// exposes methods to manage the controllerRef of deployments.
+//
+// The CanAdopt() function can be used to perform check prior to the first adoption.
+// If CanAdopt() returns a non-nil error, all adoptions will fail.
+//
+// NOTE: Once CanAdopt() is called, it will not be called again by the same
+//       DeploymentControllerRefManager instance.
+func NewDeploymentControllerRefManager(
+	client kubernetes.Interface,
+	controller metav1.Object,
+	selector labels.Selector,
+	controllerKind schema.GroupVersionKind,
+	canAdopt func() error,
+) *DeploymentControllerRefManager {
+	return &DeploymentControllerRefManager{
+		BaseControllerRefManager: basecontroller.BaseControllerRefManager{
+			Controller:   controller,
+			Selector:     selector,
+			CanAdoptFunc: canAdopt,
+		},
+		controllerKind: controllerKind,
+		client:         client,
+	}
+}
+
+// ClaimDeployments tries to take ownership of a list of Deployments.
+//
+// It will reconcile the following:
+//   * Adopt orphans if the selector matches.
+//   * Release owned objects if the selector no longer matches.
+//
+// A non-nil error is returned if some form of reconciliation was attempted and
+// failed.
+//
+// If the error is nil, either the reconciliation succeeded, or no
+// reconciliation was necessary. The list of Deployments that you now own is
+// returned.
+func (m *DeploymentControllerRefManager) ClaimDeployments(
+	deployments []*apps.Deployment) ([]*apps.Deployment, error) {
+
+	var claimed []*apps.Deployment
+	var errlist []error
+
+	match := func(obj metav1.Object) bool {
+		return m.Selector.Matches(labels.Set(obj.GetLabels()))
+	}
+	adopt := func(obj metav1.Object) error {
+		return m.adoptDeployment(obj.(*apps.Deployment))
+	}
+	release := func(obj metav1.Object) error {
+		return m.releaseDeployment(obj.(*apps.Deployment))
+	}
+
+	for _, rs := range deployments {
+		ok, err := m.ClaimObject(rs, match, adopt, release)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
+		if ok {
+			claimed = append(claimed, rs)
+		}
+	}
+
+	return claimed, utilerrors.NewAggregate(errlist)
+
+}
+
+func (m *DeploymentControllerRefManager) adoptDeployment(d *apps.Deployment) error {
```

Comment:
you should check CanAdopt first

Reply:
ClaimObject will automatically call CanAdopt this is one of the reasons we want to extract it.

Reply:
I don't see ClaimObject calling CanAdopt - would you mind pointing me there?
the other objects still check for it https://github.com/kubernetes/kubernetes/blob/6659f2a7d894a404d59265a708a2d2bd03806e98/pkg/controller/controller_ref_manager.go#L211-L214

Any reason not to keep these methods exported?

Reply:
Sorry, you right. Will fix it.